src/common/meta/src/range_stream.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use futures::future::BoxFuture;
+use futures::{ready, FutureExt, Stream};
+
+use crate::error::Result;
+use crate::kv_backend::KvBackendRef;
+use crate::rpc::store::{RangeRequest, RangeResponse};
+use crate::rpc::KeyValue;
+use crate::util::get_next_prefix_key;
+
+pub type KeyValueDecoderFn<K, V> = dyn Fn(KeyValue) -> Result<(K, V)> + Send + Sync;
+
+enum PaginationStreamState<K, V> {
+    /// At the start of reading.
+    Init,
+    /// Decoding key value pairs.
+    Decoding(SimpleKeyValueDecoder<K, V>),
+    /// Retrieving data from backend.
+    Reading(BoxFuture<'static, Result<(PaginationStreamFactory, Option<RangeResponse>)>>),
+    /// Error
+    Error,
+}
+
+pub const DEFAULT_PAGE_SIZE: usize = 512;
+
+struct PaginationStreamFactory {
+    kv: KvBackendRef,
+    /// key is the first key for the range, If range_end is not given, the
+    /// request only looks up key.
+    pub key: Vec<u8>,
```

Comment:
suggest just store the entire `RangeRequest` to make codes concise

Reply:
However, The `limit` in the `RangeRequest` isn't the `page_size` in the `PaginationStreamFactory`.